### PR TITLE
Yci/get infra state

### LINKED ISSUE #4004: editoast: create loading infra state endpoint
# Description

Create get /infra/state endpoint in editoast to get the current state of the loading of the infra. 

# Goal

We want to return the current state of an infra

# Acceptance Criteria

- [ ] openapi is updated
- [ ] the frontend uses the editoast endpoint
- [ ] The endpoint return the state of the infra

### PR DESCRIPTION
**Pull request by younesschrifi**
Close #4004

### CODE REVIEW

File: core/src/main/java/fr/sncf/osrd/api/InfraCacheStatusEndpoint.java [hunk 1]
```diff
@@ -44,8 +44,8 @@ public Response act(Request req) throws IOException {
         // Parse request input
         try {
             var body = new RqPrint(req).printBody();
-
-            if (body.equals("")) {
+            System.out.println(body);
+            if (body.equals("")|| body.equals("{\"infra\":null}")) {
```
Comment (Castavo) [inline core/src/main/java/fr/sncf/osrd/api/InfraCacheStatusEndpoint.java]: I feel like testing equality of the body against a string isn't very pretty, maybe you could do something around
```suggestion
String infra_id;
if (!body.equals("")) {
    infra_id = adapterRequest.fromJson(body).infra
}
if (infra_id == null)) {
```

File: editoast/src/views/infra/mod.rs [hunk 2]
```diff
@@ -463,14 +478,43 @@ async fn call_core_infra_load(
...
 }
 
+/// Builds a Core infra state request, runs it
+async fn call_core_cache_status(
+    payload: Option<Json<StatePayload>>,
+    db_pool: Data<DbPool>,
+    core: Data<CoreClient>,
+) -> Result<HashMap<String, InfraStateResponse>> {
+    let payload = payload.unwrap_or(Json(StatePayload { infra: None }));
+    if let Some(infra_id) = payload.infra {
```
Comment (Castavo) [inline editoast/src/views/infra/mod.rs:483]: nice destructuring, I need to use that more :grin:

File: editoast/src/views/infra/mod.rs [hunk 3]
```diff
@@ -463,14 +478,43 @@ async fn call_core_infra_load(
...
+    Ok(response)
+}
+
+#[get("/cache_status")]
+async fn cache_status(
+    payload: Option<Json<StatePayload>>,
+    db_pool: Data<DbPool>,
+    core: Data<CoreClient>,
+) -> Result<Json<HashMap<String, InfraStateResponse>>> {
+    let response = call_core_cache_status(payload, db_pool, core).await?;
```
Comment (Castavo) [inline editoast/src/views/infra/mod.rs]: Why move the logic out of the endpoint function if it leaves only two lines in the endpoint function ?

Comment (younesschrifi) [reply editoast/src/views/infra/mod.rs]: You are right (me = no brain)

File: core/src/main/java/fr/sncf/osrd/api/InfraCacheStatusEndpoint.java [hunk 4]
```diff
@@ -44,15 +44,23 @@ public Response act(Request req) throws IOException {
...
+                if (infra == null) {
+                    infraManager.forEach((infraId, infraCacheEntry) -> {
+                        res.put(infraId, SerializedInfraCache.from(infraCacheEntry));
+                    });
+                } else {
+                    var request = adapterRequest.fromJson(body);
+                    var infraCacheEntry = infraManager.getInfraCache(request.infra);
+                    res.put(request.infra, SerializedInfraCache.from(infraCacheEntry));
+                }
+            } else {
```
Comment (Castavo) [inline core/src/main/java/fr/sncf/osrd/api/InfraCacheStatusEndpoint.java]: You could avoid the redundancy by doing
```suggestion
            }
            if (infra != null) {
                var request = adapterRequest.fromJson(body);
                var infraCacheEntry = infraManager.getInfraCache(request.infra);
                res.put(request.infra, SerializedInfraCache.from(infraCacheEntry));
            } else {
```

**Review by flomonster — APPROVED** [review]
LGTM but didn't test it.

File: core/src/main/java/fr/sncf/osrd/api/InfraCacheStatusEndpoint.java [hunk 5]
```diff
@@ -44,15 +44,18 @@ public Response act(Request req) throws IOException {
         // Parse request input
         try {
             var body = new RqPrint(req).printBody();
-
-            if (body.equals("")) {
+            String infra;
```
Comment (Khoyo) [inline core/src/main/java/fr/sncf/osrd/api/InfraCacheStatusEndpoint.java]: Why is this declared here?

File: core/src/main/java/fr/sncf/osrd/api/InfraCacheStatusEndpoint.java [hunk 6]
```diff
@@ -44,15 +44,18 @@ public Response act(Request req) throws IOException {
...
+            if (!body.equals("")) {
+                infra = adapterRequest.fromJson(body).infra;
+                if (infra != null) {
+                    var request = adapterRequest.fromJson(body);
```
Comment (Khoyo) [inline core/src/main/java/fr/sncf/osrd/api/InfraCacheStatusEndpoint.java]: Can't we parse that once, then check is `request.infra` is null?

Additionally, we should probably return an error if it is.

**Review by Khoyo — APPROVED** [review]
Not tested, but the core part LGTM

**Pull request merged**